### ISSUE TITLE
Critical bug: Cluster jobs do NOT work

### ISSUE THREAD

**Issue by asoplata** [DynaSim/DynaSim#181]
I was beginning to test the benchmark code on the cluster (using the current `debug` branch), but I can't get my cluster jobs to work AT ALL, due to some new bug in how DynaSim handles batch submission.

### Steps to reproduce

1. Make new `batchdirs` folder from scratch in home.
2. Clone DynaSim, checkout `debug` branch
3. Run the following script, which uses only the default installed mechanisms:

```{matlab}
run_name = 'cluster_bug_test';

% Define equations of cell model (same for all populations)
eqns={
  'dv/dt=Iapp+@current';
};

time_end = 500; % in milliseconds
numEcells = 20;
numIcells = 8;

% Create DynaSim specification structure
s=[];
s.populations(1).name='E';
s.populations(1).size=numEcells;
s.populations(1).equations=eqns;
s.populations(1).mechanism_list={'iNa','iK','ileak'};
s.populations(1).parameters={'Iapp',0};

s.populations(2).name='I';
s.populations(2).size=numIcells;
s.populations(2).equations=eqns;
s.populations(2).mechanism_list={'iNa','iK','ileak'};
s.populations(2).parameters={'Iapp',0};

s.connections(1).direction='E->I';
s.connections(1).mechanism_list={'iAMPA'};
s.connections(2).direction='I->E';
s.connections(2).mechanism_list={'iGABAa'};
s.connections(3).direction='E->E';
s.connections(3).mechanism_list={'iAMPA'};
s.connections(4).direction='I->I';
s.connections(4).mechanism_list={'iGABAa'};

vary={
  '(E)',           'Iapp',     [0.3,1.3];
};

%% Set simulation parameters
% How much RAM, options: 8G?, 24, 48, 96, 128
memory_limit = '8G';
% memory_limit = '16G';
% memory_limit = '48G';
% memory_limit = '96G';
% memory_limit = '254G';

% Flags
% Save data/results to this directory. If just a single name, will
%   save to that directory name in the current directory from which it's run.
%   Will create directory if it does not exist.
data_dir = run_name;
% data_dir = strcat('/projectnb/crc-nak/asoplata/x7-scc-data/',...
%                   run_name);
save_results_flag = 1;
save_data_flag =    1;
cluster_flag =      1;
% overwrite_flag =    1;
% % Even if `save_data_flag` is 0, if running on cluster this must be off too in
% %   order to not save data?
verbose_flag =      1;
% compile_flag =      0;
% disk_flag =         0;
% downsample_factor = 1;

% local run of the simulation,
%   i.e. in the interactive session you're running this same script in
data = dsSimulate(s,'study_dir',data_dir,...
    'save_results_flag',save_results_flag,'save_data_flag',save_data_flag,...
    'cluster_flag',cluster_flag,'tspan',[0 time_end],...
    'verbose_flag',verbose_flag,...
    'vary',vary);
                  % 'plot_functions',{@dsPlot,@dsPlot,@dsPlot},...
                  % 'plot_options',{{'plot_type','waveform','format','png'},...
                  %                 {'plot_type','power','format','png'},...
                  %                 {'plot_type','rastergram','format','png'}});


% data = dsSimulate(s,'save_data_flag',save_data_flag,'study_dir',data_dir,...
%                   'cluster_flag',cluster_flag,'verbose_flag',verbose_flag,...
%                   'overwrite_flag',overwrite_flag,'tspan',[0 time_end],...
%                   'save_results_flag',save_results_flag,'solver','euler',...
%                   'memory_limit',memory_limit,'compile_flag',compile_flag,...
%                   'disk_flag',disk_flag,'downsample_factor',downsample_factor,...
%                   'vary',vary,...
%                   'plot_functions',{@dsPlot,@dsPlot,@dsPlot},...
%                   'plot_options',{{'plot_type','waveform','format','png'},...
%                                   {'plot_type','power','format','png'},...
%                                   {'plot_type','rastergram','format','png'}});
% exit
```

4. Observe the following error output:

```{matlab}
>> run cluster_bug_test
PREPARING STUDY:
Creating study directory: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test
Creating data directory: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/data
created temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531135902416_0
saving study file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/studyinfo.mat
deleting temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531135902416_0

PREPARING SOLVER:
Creating solver directory /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve
Changing directory to /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve
Saving params.mat
Creating solver file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve/solve_ode_cluster_bug_test.m
Changing directory back to /usr3/graduate/asoplata/s1-dynasim/dynasim

PREPARING BATCH:
Creating 2 cluster jobs in batch directory for 2 simulations...
Creating file listing jobs in batch directory: /usr3/graduate/asoplata/batchdirs/scriptlist.txt
Creating distinct solver sub-directories for 2 simulations...
created temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531135904370_0
saving study file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/studyinfo.mat
deleting temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531135904370_0
Submitting cluster jobs with shell command: "qmatjobs_memlimit_loop batchdirs 8G"
Submit command failed: "qmatjobs_memlimit_loop batchdirs 8G"
/usr3/graduate/asoplata/batchdirs/batchdirs: No such file or directory.

>> 
```

It seems to be getting doubly confused:

1. The study directory and solve file are fine, but
2. It doesn't create a subfolder in `batchdirs` from the name of the study, instead choosing to just throw up the files used for the batch job (e.g. `scriptlist.txt`, `sim_job1.m`, etc.) directly into `~/batchdirs` itself,
3. It finally gets confused and seems to accidentally think `~/batchdirs/batchdirs` is what it's looking for, AFTER it's done this.

If I DO create a `~/batchdirs/batchdirs` subfolder, then it still doesn't work, even though it claims to submit the jobs (this does NOT work):

```{matlab}
>> run cluster_bug_test
PREPARING STUDY:
Creating study directory: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test
Creating data directory: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/data
created temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531140955302_0
saving study file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/studyinfo.mat
deleting temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531140955302_0

PREPARING SOLVER:
Creating solver directory /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve
Changing directory to /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve
Saving params.mat
Creating solver file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/solve/solve_ode_cluster_bug_test.m
Changing directory back to /usr3/graduate/asoplata/s1-dynasim/dynasim

PREPARING BATCH:
Creating 2 cluster jobs in batch directory for 2 simulations...
Creating file listing jobs in batch directory: /usr3/graduate/asoplata/batchdirs/scriptlist.txt
Creating distinct solver sub-directories for 2 simulations...
created temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531140957407_0
saving study file: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/studyinfo.mat
deleting temporary lock file for this process: /usr3/graduate/asoplata/s1-dynasim/dynasim/cluster_bug_test/.lock_20170531140957407_0
Submitting cluster jobs with shell command: "qmatjobs_memlimit_loop batchdirs 8G"
Submit command status: 
cat: scriptlist.txt: No such file or directory

2 jobs successfully submitted!
>> 
```

As far as I can tell, this means the the current debug branch is broken w.r.t. running jobs on the cluster. Has anyone done so using the most up to date `debug` branch?

**Comment by asoplata** [comment]
Who was the last person to work on the batch submission code? @erik-roberts ?

**Comment by asoplata** [comment]
So above, I was testing the current `debug` branch, which we expect to be buggy. However, I now also CANNOT get this to work on the current `master` either, as it gives a different error.

If you run the same script (except with `SimulateModel` in place of `dsSimulate`) below:

```{matlab}
run_name = 'cluster_bug_test';

% Define equations of cell model (same for all populations)
eqns={
  'dv/dt=Iapp+@current';
};

time_end = 500; % in milliseconds
numEcells = 20;
numIcells = 8;

% Create DynaSim specification structure
s=[];
s.populations(1).name='E';
s.populations(1).size=numEcells;
s.populations(1).equations=eqns;
s.populations(1).mechanism_list={'iNa','iK','ileak'};
s.populations(1).parameters={'Iapp',0};

s.populations(2).name='I';
s.populations(2).size=numIcells;
s.populations(2).equations=eqns;
s.populations(2).mechanism_list={'iNa','iK','ileak'};
s.populations(2).parameters={'Iapp',0};

s.connections(1).direction='E->I';
s.connections(1).mechanism_list={'iAMPA'};
s.connections(2).direction='I->E';
s.connections(2).mechanism_list={'iGABAa'};
s.connections(3).direction='E->E';
s.connections(3).mechanism_list={'iAMPA'};
s.connections(4).direction='I->I';
s.connections(4).mechanism_list={'iGABAa'};

vary={
  '(E)',           'Iapp',     [0.3,1.3];
};

%% Set simulation parameters
% How much RAM, options: 8G?, 24, 48, 96, 128
memory_limit = '8G';
% memory_limit = '16G';
% memory_limit = '48G';
% memory_limit = '96G';
% memory_limit = '254G';

% Flags
% Save data/results to this directory. If just a single name, will
%   save to that directory name in the current directory from which it's run.
%   Will create directory if it does not exist.
data_dir = run_name;
% data_dir = strcat('/projectnb/crc-nak/asoplata/x7-scc-data/',...
%                   run_name);
save_results_flag = 1;
save_data_flag =    1;
cluster_flag =      1;
% overwrite_flag =    1;
% % Even if `save_data_flag` is 0, if running on cluster this must be off too in
% %   order to not save data?
verbose_flag =      1;
% compile_flag =      0;
% disk_flag =         0;
% downsample_factor = 1;

% local run of the simulation,
%   i.e. in the interactive session you're running this same script in
data = SimulateModel(s,'study_dir',data_dir,...
    'save_results_flag',save_results_flag,'save_data_flag',save_data_flag,...
    'cluster_flag',cluster_flag,'tspan',[0 time_end],...
    'verbose_flag',verbose_flag,...
    'vary',vary);
                  % 'plot_functions',{@dsPlot,@dsPlot,@dsPlot},...
                  % 'plot_options',{{'plot_type','waveform','format','png'},...
                  %                 {'plot_type','power','format','png'},...
                  %                 {'plot_type','rastergram','format','png'}});


% data = dsSimulate(s,'save_data_flag',save_data_flag,'study_dir',data_dir,...
%                   'cluster_flag',cluster_flag,'verbose_flag',verbose_flag,...
%                   'overwrite_flag',overwrite_flag,'tspan',[0 time_end],...
%                   'save_results_flag',save_results_flag,'solver','euler',...
%                   'memory_limit',memory_limit,'compile_flag',compile_flag,...
%                   'disk_flag',disk_flag,'downsample_factor',downsample_factor,...
%                   'vary',vary,...
%                   'plot_functions',{@dsPlot,@dsPlot,@dsPlot},...
%                   'plot_options',{{'plot_type','waveform','format','png'},...
%                                   {'plot_type','power','format','png'},...
%                                   {'plot_type','rastergram','format','png'}});
% exit
```

then you'll get an error in `~/batchdirs/cluster_bug_test/pbsout/sim_job1.err` about an invalid character. This error, which is different from the one above, is because the function name in `~/batchdirs/cluster_bug_test/sim_job1.m` leads to the PATH of the `sim_job1.m` file, NOT a usable function name like `sim_job1`.

Note that these errors may be compounded in `debug`, and so there may be more than one error preventing `cluster_flag` usage right now. As it stands, I can't get `debug` or `master` to work.

**Comment by benpolletta** [comment]
Hey Austin,

FWIW, I definitely ran jobs on the cluster with both dev (as of some time ago) and master in its current state. So, there must be something special going on in this example. For instance, the demos work, including cluster submission, last commit I checked.

Not that we shouldn't fix the issue.

B

**Comment by asoplata** [comment]
Update: I just checked the first cluster example from `demos/tutorial.m` on `master` and I'm still getting the runtime error related to `sim_job1.m`. I ran the following:

```{matlab}
study_dir='study_HH_varyI_cluster'; 
eqns='dv/dt=@current+I; {iNa,iK}';
vary={'','I',[0 10 20]};

[data,studyinfo]=SimulateModel(eqns,'vary',vary,...
  'study_dir',study_dir,'save_data_flag',1,'cluster_flag',1,'verbose_flag',1);
```

So apparently the cluster code in the tutorial doesn't work either? Note that code is being EXECUTED by the batch system, using the cluster, in parallel -- just that the `sim_job1.m` etc. code has a function name error, and possibly others.

**Comment by asoplata** [comment]
@benpolletta, when was the last time you tested `master`?

**Comment by erik-roberts** [comment]
fixed the master bug. i had accidentally used the wrong variable name for the function name in fprintf.

**Comment by erik-roberts** [comment]
fixed debug bug.  there was an issue with the way fileparts2 was being used.  i had initially done a batch find and replace of fileparts to fileparts2. the only issue is that fileparts was sometimes being used to find the terminal directory, instead of the parts of a file, which was not a use case i expected.  this is an example of one of those situations.

**Comment by erik-roberts** [comment]
new error now:
```
command =  sim_job1
cwd =  /projectnb/crc-nak/erob/batchdirs/demo_cluster_1
HOSTNAME =  scc-na8.scc.bu.edu
JOB_NAME =  9211393
JOB_NAME =  demo_cluster_1_sim_job1

                            < M A T L A B (R) >
                  Copyright 1984-2014 The MathWorks, Inc.
                    R2014a (8.3.0.532) 64-bit (glnxa64)
                             February 11, 2014

 
To get started, type one of these: helpwin, helpdesk, or demo.
For product information, visit www.mathworks.com.
 

SimIDs =

     1

Error: Matrix dimensions must agree.
	 in updateStudy (line 51)
	 in sim_job1 (line 11)
finished
```

**Comment by erik-roberts** [comment]
fix 2118c9b

**Comment by benpolletta** [comment]
Version issue: cluster job submission fails with 2017a. Here is the error:

PREPARING BATCH:
created temporary lock file for this process: /projectnb/crc-nak/brpp/model-dnsim-kramer_IB/kramer_IB/kramer_IB_15_40_2.488/.lock_20170608154040071_0
saving study file: /projectnb/crc-nak/brpp/model-dnsim-kramer_IB/kramer_IB/kramer_IB_15_40_2.488/studyinfo.mat
deleting temporary lock file for this process: /projectnb/crc-nak/brpp/model-dnsim-kramer_IB/kramer_IB/kramer_IB_15_40_2.488/.lock_20170608154040071_0
Submitting cluster jobs with shell command: "echo '/projectnb/crc-nak/brpp/DynaSim/functions/+ds/qmatjob_array_one_file /usr2/postdoc/brpp/batchdirs sim_job' | qsub -V -hard -l 'h_vmem=8G' -wd /usr2/postdoc/brpp/batchdirs -N batchdirs_sim_job -t 1-41:1"
Submit command failed: "echo '/projectnb/crc-nak/brpp/DynaSim/functions/+ds/qmatjob_array_one_file /usr2/postdoc/brpp/batchdirs sim_job' | qsub -V -hard -l 'h_vmem=8G' -wd /usr2/postdoc/brpp/batchdirs -N batchdirs_sim_job -t 1-41:1"
qsub: /lib64/libz.so.1: version `ZLIB_1.2.3.3' not found (required by /share/pkg/matlab/2017a/install/bin/glnxa64/libxml2.so.2)

**Comment by asoplata** [comment]
Ben is your simulation using the compile functionality? I see some Google
hits for MEX compilation issues with this library. If it is, see if you get
the issue without using compilation​.

Regardless, that looks like a MATLAB dependency bug, either by MATLAB
itself or the install on the actual cluster! Impressive.

**Comment by erik-roberts** [comment]
we should tell bu:
https://www.mathworks.com/matlabcentral/newsreader/view_thread/140611

for the time being, i'd switch back to 2016b.

**Comment by asoplata** [comment]
The un-compiled test simulations I mentioned above now seem to work properly on `debug` and `master`. I'm closing this because Erik fixed the cluster submission bugs. We should move the compiler bugs to a new issue, instead of discussing it here.
